src/calendar/export/CalendarImporterDialog.ts

```diff
@@ -9,97 +9,145 @@ import { lang } from "../../misc/LanguageViewModel"
+ * will assign event id according to the calendarGroupRoot and the long/short event status */
+export function sortOutParsedEvents(
+	parsedEvents: ParsedEvent[],
+	existingEvents: Array<CalendarEvent>,
+	calendarGroupRoot: CalendarGroupRoot,
+	zone: string,
+): {
+	rejectedEvents: RejectedEvents
+	eventsForCreation: Array<EventWrapper>
+} {
 	const instanceIdentifierToEventMap = new Map()
 	for (const existingEvent of existingEvents) {
 		if (existingEvent.uid == null) continue
 		instanceIdentifierToEventMap.set(makeInstanceIdentifier(existingEvent), existingEvent)
 	}
-	const flatParsedEvents = parsedEvents.flat()
+
 	const rejectedEvents: RejectedEvents = new Map()
-	// Don't try to create event which we already have
 	const eventsForCreation: Array<{ event: CalendarEvent; alarms: Array<AlarmInfo> }> = []
-	for (const { event, alarms } of flatParsedEvents) {
-		const rejectionReason = shouldBeSkipped(event, instanceIdentifierToEventMap)
-		if (rejectionReason != null) {
-			getFromMap(rejectedEvents, rejectionReason, () => []).push(event)
-			continue
-		}
-
-		// hashedUid will be set later in calendarFacade to avoid importing the hash function here
-		const repeatRule = event.repeatRule
-		event._ownerGroup = calendarGroupRoot._id
-
-		if (repeatRule && repeatRule.timeZone === "") {
-			repeatRule.timeZone = getTimeZone()
-		}
-
-		for (let alarmInfo of alarms) {
-			alarmInfo.alarmIdentifier = generateEventElementId(Date.now())
+	for (const [_, flatParsedEvents] of groupBy(parsedEvents, (e) => e.event.uid)) {
+		let progenitor: { event: CalendarEvent; alarms: Array<AlarmInfo> } | null = null
+		let alteredInstances: Array<{ event: CalendarEvent; alarms: Array<AlarmInfo> }> = []
+
+		for (const { event, alarms } of flatParsedEvents) {
+			const rejectionReason = shouldBeSkipped(event, instanceIdentifierToEventMap)
+			if (rejectionReason != null) {
+				getFromMap(rejectedEvents, rejectionReason, () => []).push(event)
+				continue
+			}
+
+			// hashedUid will be set later in calendarFacade to avoid importing the hash function here
+			const repeatRule = event.repeatRule
+			event._ownerGroup = calendarGroupRoot._id
+
+			if (repeatRule != null && repeatRule.timeZone === "") {
+				repeatRule.timeZone = getTimeZone()
+			}
+
+			for (let alarmInfo of alarms) {
+				alarmInfo.alarmIdentifier = generateEventElementId(Date.now())
+			}
+
+			assignEventId(event, zone, calendarGroupRoot)
+			if (event.recurrenceId == null) {
+				// the progenitor must be null here since we would have
+				// rejected the second uid-progenitor event in shouldBeSkipped.
+				progenitor = { event, alarms }
+			} else {
+				if (progenitor?.event.repeatRule != null) {
+					insertIntoSortedArray(
+						createDateWrapper({ date: event.recurrenceId }),
+						progenitor.event.repeatRule.excludedDates,
+						(left, right) => left.date.getTime() - right.date.getTime(),
+						() => true,
+					)
+				}
+				alteredInstances.push({ event, alarms })
```

Comment:
should we still create it if there's no progenitor? we probably should but we should make sure it doesn't invalidate some invariant somewhere

Reply:
yes, it's valid to have no knowledge of the progenitor since we are allowed to send invitations to only an altered instance of a series.